### DISCUSSION TITLE
cmake + vcpkg : how to verify platform toolset for VC/Windows builds?

### DISCUSSION THREAD

**Discussion by jeff-kohn** [microsoft/vcpkg#29898] [Q&A]
I have a cmake project that targets both windows and linux. For the windows build, I'm specifying the v142 platform toolset using VCPKG_TARGET_TRIPLET, with a custom triplet that specifies the appropriate VCPKG_PLATFORM_TOOLSET, because the users of my library are still building with VS2019 and v142 toolset.

This is my custom triplet:
```
set(VCPKG_TARGET_ARCHITECTURE x86)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE static)
set(VCPKG_PLATFORM_TOOLSET "v142")
```

It's getting passed to vcpkg via cmake by setting the VCPKG_TARGET_TRIPLET cache variable equal to the name of my custom triplet. (yes, I deleted the cmake cache and ran configure before building again).

Everything in my project builds fine, and vcpkg output shows that it's using my triplet. When attempting to link my static library into the application using it, it works for Debug builds; but release builds are failing with the following error:

```
1>LINK : fatal error C1007: unrecognized flag '-Zc:nrvo' in 'p2'
1>LINK : fatal error LNK1257: code generation failed

```

I'm not explicitly setting this flag, but according to [this documentation](https://learn.microsoft.com/en-us/cpp/build/reference/zc-nrvo?view=msvc-170) it's auto-enabled in the latest toolset versions when using permissive- (which I am using). This particular linker flag is only supported in VS2022, version 17.4 or later (v143 toolset). So I can only conclude that somehow my library is getting built with the wrong toolset.

Any thoughts on how to troubleshoot/fix this? I've tried passing --debug to vcpkg to troubleshoot, but when I look in vcpkg-manifest-install.log I don't see anything that would indicate what's going on. The v143 toolset only shows up in the logfile in relation to the host triplet vcpkg uses for compiler detection stuff. When building targets it appears to be using my custom triplet. I even tried specifying my custom triplet for the VCPKG_HOST_TRIPLET var, but that didn't seem to make any difference in the error I'm getting (although it did remove the reference to v143 toolset that was showing up in the log).

**Comment by ccptoebeans** [comment]
I am having the same problem. Using using vcpkg with Ninja Multi-Config. Is nobody else having this issue?

**Reply by Neumann-A** [reply]
Is the error in your CMake project or within vpckg itself? I assume it is in your cmake project and then the problem is that you are running from the wrong vs dev promt which isn't really a vcpkg problem.

**Reply by jeff-kohn** [reply]
The problem is an inconsistency in vcpkg. The toolset value in the triplet only applies to the dependencies vcpkg builds, not your actual project. This is extremely unintuitive because the VCPKG toolchain file is also being used to build the project, and VCPKG knows how to create a shell environment with the needed build settings, so why doesn't it? 

I wrote a little more about the answers I found for myself over here: https://stackoverflow.com/questions/75575785/cmake-vcpkg-how-to-verify-platform-toolset-for-vc-windows-builds/75637196#75637196

Short answer if you're using VS or VSCode to build is to add this to your configure preset (btw this is for Ninja, Visual Studio generator may use a slightly different property syntax from what I recall):
```
         "toolset": {
            "value": "version=14.29",
            "strategy": "external"
         },
```
         
  For a command line you have to make sure you're running from a Developer Command Prompt with the correct environment loaded. Just keep in mind if you're not careful you could end up with VCPKG using a different toolset for your dependencies and your main build.

**Reply by Neumann-A** [reply]
> This is extremely unintuitive because the VCPKG toolchain file is also being used to build the project

since when? vcpkg.cmake is just a script helper to setup some cmake paths and install dependencies (only manifest mode) it does not handle the dev environment of the project itself. 

> VCPKG knows how to create a shell environment with the needed build settings

It doesn't. Just on windows it assume you want to use VS somehow. 

> not careful you could end up with VCPKG using a different toolset for your dependencies and your main build

see above it does not know what you want. It just tries to have a best guess about it.

Why can it not handle that? The reason is simple, it simply cannot change the parents environment.

So you start cmake (configure) in one environment. CMake calls vcpkg. vcpkg setups its own env on windows but cannot change what cmake sees. Even if it could, if you execute cmake --build or similar that again runs in the top level environment vcpkg has no leverage to change.

**Reply by jeff-kohn** [reply]
VCPKG **does** know how to set up a build environment on Windows, it does so whenever you run "vcpkg install" or invoke vcpkg in manifest mode from a build. I don't have to run from a Developer Command Prompt for this to work, I can open up a regular command prompt  and just run "vcpkg install openssl" and it will find the compiler and build everything for me. 

What's more, according to [this page]( https://learn.microsoft.com/en-us/vcpkg/users/triplets) the VCPKG_PLATFORM_TOOLSET variable "Specifies the Visual Studio-based C/C++ compiler toolchain to use."   That's a direct quote, so if that's not what this variable is supposed to do it should really be documented better.

What I have observed is the following:
1) Set up a cmake project to use vcpkg in manifest mode, and specify that cmake should use the vcpkg toolchain 
2) Set up a custom triplet that specifies v142 toolset using VCPKG_PLATFORM_TOOLSET 
3) Build

What I've found is that if you follow those steps, you'll get a build where your dependences are built with v142 while your project is built with the latest installed toolset. With all due respect, nothing you can say will convince me that this is the correct and desirable behavior.

Either the VCPKG_PLATFORM_TOOLSET variable is broken, or it's a poorly documented footgun that should be removed altogether because it accomplishes nothing except to cause inconsistent builds.

As for the rest of your argument, CMake doesn't just "call" vcpkg, it's using a toolchain file provided by vcpkg. That toolchain is supposed to be responsible for setting up the correct variables etc to build. If I'm using the same toolchain file for building my project that is being used to build my dependencies, it's pretty reasonable to expect I should get the same build environment for both.

**Reply by Neumann-A** [reply]
> That toolchain is supposed to be responsible for setting up the correct variables etc to build.

Yeah that what a toolchain normally does but in vcpkg case it is just a code injection variable to make stuff build by vcpkg findable for your project. You are supposed to provide your own toolchain via VCPKG_CHAINLOAD_TOOLCHAIN_FILE or have a different approach how you invoke/inject vcpkg. The reason why you set CMAKE_TOOLCHAIN_FILE to vcpkg.cmake is that, that script has to run fairly early in the cmake configure stage. You could also use CMAKE_PROJECT_TOP_LEVEL_INCLUDES or CMAKE_PROJECT_INCLUDE or even a simple include() statement. 

> That's a direct quote, so if that's not what this variable is supposed to do it should really be documented better.

Maybe read that in the context of: "within/internal to vcpkg". Normal CMake rules still apply to the top level project since that is outside of vcpkg....

> With all due respect, nothing you can say will convince me that this is the correct and desirable behavior.

Why not? That is a totally supported use case in VS. Old MSVC stuff can be linked with newer MSVC libs as long as the toplevel project uses the newest linker?

> I don't have to run from a Developer Command Prompt for this to work, I can open up a regular command prompt and just run "vcpkg install openssl" and it will find the compiler and build everything for me.

Then write a portfile and do `vcpkg install yourproject`. Then the project gets build from within vcpkg with everything correctly setup?
